okhttp/src/main/java/io/grpc/okhttp/OkHttpServerTransport.java

```diff
@@ -161,10 +161,12 @@ public void start(ServerTransportListener listener) {
 
   private void startIo(SerializingExecutor serializingExecutor) {
     try {
-      bareSocket.setTcpNoDelay(true);
       HandshakerSocketFactory.HandshakeResult result =
-          config.handshakerSocketFactory.handshake(bareSocket, Attributes.EMPTY);
-      Socket socket = result.socket;
+          config.handshakerSocketFactory.handshake(socket, Attributes.EMPTY);
+      socket.setTcpNoDelay(true);
```

Comment:
Move this back before handshake()? Given we talked about this, I assume something didn't like it.

Reply:
That TSAN failure looks like a bug in Java's Socket. Glancing at the JDK, [it is synchronized](https://github.com/openjdk/jdk/blob/bfef3c3e80d4df98bdb0d853c2e2711cdd8252ce/src/java.base/share/classes/java/net/Socket.java#L586). But that was a recent change. It was busted really recently, as seen with blame. https://bugs.openjdk.org/browse/JDK-8278326

So that failure looks like it is already fixed upstream; it just needs to sync internally. But it also looks like all previous versions of Socket were broken. This makes shutdown overall broken. Eww...

Let's move the `setTcpNoDelay(true)` up before handshake(), but put it in a `synchronized`. That will then either 1) initializes the socket, and guarantee it doesn't run concurrently with shutdown() or 2) throw, because shutdown() already closed the socket. And we should have a comment linking to that JDK issue.

Looking back at https://github.com/grpc/grpc-java/issues/10228 , the same problem was happening there. Could you run a test with `this.socket = result.socket;` commented out to see if the "synchronized setTcpNoDelay" fixes the problem by itself? Commenting out `this.socket = result.socket;` is the equivalent of swapping back to the "final bareSocket". It'd take some work for me to find the TLS race I saw before, and I wonder if the problem was actually just always this Socket bug.

Reply:
Yes, Keeping the socket.setTcpNoDelay(true) before the handshake results in other [data race errors](https://fusion2.corp.google.com/invocations/d69c6b7c-9c0d-4628-894c-0883268fad13/targets/%2F%2Fthird_party%2Fjava_src%2Fgrpc%2Fokhttp:src%2Ftest%2Fjava%2Fio%2Fgrpc%2Fokhttp%2FOkHttpServerTransportTest/tests)

Reply:
Neither seems to fix the issue. 

Commenting out `this.socket = result.socket` with a synchronized lock to the setTcpNoDelay (moved back up) gives us [112/1000 failures](https://fusion2.corp.google.com/invocations/e2ac6f0e-1994-4da9-8e9b-151782b45fa8/targets/%2F%2Fthird_party%2Fjava_src%2Fgrpc%2Fokhttp:src%2Ftest%2Fjava%2Fio%2Fgrpc%2Fokhttp%2FOkHttpServerTransportTest).

And keeping that line in with the synchronized lock to the setTcpNoDelay (moved back up) gives us [127/1000 failures](https://fusion2.corp.google.com/invocations/f97537e2-158e-4d5d-8ba3-804ebdbb8017/targets/%2F%2Fthird_party%2Fjava_src%2Fgrpc%2Fokhttp:src%2Ftest%2Fjava%2Fio%2Fgrpc%2Fokhttp%2FOkHttpServerTransportTest).

It seems like the original code with the bareSocket (as well as my revised one) passes all test cases when the `setTcpNoDelay(true)` is either moved down or removed entirely. A synchronized lock on `setTcpNoDelay(true)` seems to bring the failures down from ~300/1000 to ~100/1000.

Reply:
> And keeping that line in with the synchronized lock to the setTcpNoDelay (moved back up) gives us 127/1000 failures.

Ugh. I'm realizing that failure is a bug in the test. The race isn't with shutdown there. The "client" (in the test) calls close and it is considering that a race. In real life that would come from the kernel. What's even worse is PipeSocket doesn't actually use any Socket resources; it is a fake. We'll probably need to replace that with a real socket. That also invalidates some of the testing you've done up to this point.

Try to reproduce a failure using https://github.com/grpc/grpc-java/pull/10281 . Assuming that makes things look better, I can send that out as a PR. See go/grpc-prod/java/importing#cherry-picking-prs for how to copy it to your client.

We first noticed the TSAN failure when I saw a TLS handshake failure. Later, Ivy created the issue she assigned to you with a recent failure, but it was a different failure than what I saw. The failure in that issue may be a test problem, and so assuming my change fixes that test problem you'll want to run some more tests to see if you can trigger other failures. (We can chat about this over GVC)

> It seems like the original code with the bareSocket (as well as my revised one) passes all test cases when the setTcpNoDelay(true) is either moved down or removed entirely.

That's probably because handshake() does very little in these tests.

Reply:
We've been barking up the wrong tree. #10281 will fix the TSAN failure you were looking at. The problem there shouldn't be fixed by changing the OkHttpServerTransport. However, I reproduced https://github.com/grpc/grpc-java/issues/10284 and _that_ issue _does_ need a fix like we have here.